interfaces/builtin/kernel_module_load.go

```diff
@@ -0,0 +1,220 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/kmod"
+	"github.com/snapcore/snapd/snap"
+)
+
+const kernelModuleLoadSummary = `allows to statically control kernel module loading`
+
+const kernelModuleLoadBaseDeclarationPlugs = `
+  kernel-module-load:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const kernelModuleLoadBaseDeclarationSlots = `
+  kernel-module-load:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-connection: true
+`
+
+var modulesAttrTypeError = errors.New(`kernel-module-load "modules" attribute must be a list of dictionaries`)
+
+// kernelModuleLoadInterface allows creating transient and persistent modules
+type kernelModuleLoadInterface struct {
+	commonInterface
+}
+
+type loadOption int
+
+const (
+	loadNone loadOption = iota
+	loadDenied
+	loadOnBoot
+)
+
+type ModuleInfo struct {
+	name    string
+	load    loadOption
+	options string
+}
+
+var kernelModuleNameRegexp = regexp.MustCompile(`^[a-zA-Z0-9_]+$`)
+
+func enumerateModules(plug interfaces.Attrer, handleModule func(moduleInfo *ModuleInfo) error) error {
+	modulesAttr, ok := plug.Lookup("modules")
+	if !ok {
+		return nil
+	}
+	modules, ok := modulesAttr.([]interface{})
+	if !ok {
+		return modulesAttrTypeError
+	}
+
+	for _, m := range modules {
+		module, ok := m.(map[string]interface{})
+		if !ok {
+			return modulesAttrTypeError
+		}
+
+		name, ok := module["name"].(string)
+		if !ok {
+			return errors.New(`kernel-module-load "name" must be a string`)
+		}
+
+		var load loadOption
+		if loadAttr, found := module["load"]; found {
+			loadString, ok := loadAttr.(string)
+			if !ok {
+				return errors.New(`kernel-module-load "load" must be a string`)
+			}
+
+			switch loadString {
+			case "denied":
+				load = loadDenied
+			case "on-boot":
+				load = loadOnBoot
+			default:
+				return fmt.Errorf(`kernel-module-load "load" value is unrecognized: %q`, loadString)
+			}
+		}
+
+		var options string
+		if optionsAttr, found := module["options"]; found {
+			options, ok = optionsAttr.(string)
+			if !ok {
+				return errors.New(`kernel-module-load "options" must be a string`)
+			}
+		}
+
+		moduleInfo := &ModuleInfo{
+			name:    name,
+			load:    load,
+			options: options,
+		}
+
+		if err := handleModule(moduleInfo); err != nil {
+			return err
+		}
```

Comment:
Would it make sense to have a separate loop that iterates over all moduleInfos and calls handleModule for them? That way we would first enumerate them, and only if all modules were fine we would handle them?

Reply:
I don't think so, because `handleModule()` itself is a validation function (if you see how BeforeConnectPlug() uses enumerateModules()), so you wouldn't gain much.